src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1818,6 +1818,19 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                     int right_kind = ASRUtils::extract_kind_from_ttype_t(right_type);
                     bool is_left_f32 = ASR::is_a<ASR::Real_t>(*left_type) && left_kind == 4;
                     bool is_right_f32 = ASR::is_a<ASR::Real_t>(*right_type) && right_kind == 4;
+                    if((left_kind != right_kind) || 
```

Comment:
Also one thing to checkout here that is integer division by `//` still allows compatibility between these types on this branch, which I'm not sure should be allowed or not. Maybe @certik can tell if that should be allowed.
On face value I think it should not be allowed, as you can see these different types are not overloaded in the `lpython_floordiv() `function too.
```
@overload
def _lpython_floordiv(a: f64, b: f64) -> f64:
@overload
def _lpython_floordiv(a: f32, b: f32) -> f32:
@overload
def _lpython_floordiv(a: i8, b: i8) -> i8:
@overload
def _lpython_floordiv(a: i16, b: i16) -> i16:
@overload
def _lpython_floordiv(a: i32, b: i32) -> i32: 
@overload
def _lpython_floordiv(a: i64, b: i64) -> i64:
@overload
def _lpython_floordiv(a: bool, b: bool) -> bool:
```

Reply:
I think `//` should also be required to have the same type.

Reply:
Done!